File: leetcode/LinkedListTests/LinkedList_328.cs
namespace LinkedListTests;

[TestFixture]
internal class LinkedList_328
{
    class Solution {
        public ListNode OddEvenList(ListNode head) {
            if(head is null) return head;
            var odd = head;
            var even = head.next;
            var evenHead = even;
        
            while(even is not null && even.next is not null) {
                odd.next = even.next;
                odd = odd.next;
                even.next = odd.next;
                even = even.next;
            }
            odd.next = evenHead;
            return head;
        }
    }
}

File: leetcodePython/LinkedList/linkedlist_328.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, nxt=None):
        self.val = val
        self.next = nxt


class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None
        prevEven = evenHead = ListNode()
        prevOdd = oddHead = ListNode()
        while head:
            odd = head
            even = head.next
            odd.next = None
            prevOdd.next = odd
            prevEven.next = even
            prevOdd = prevOdd.next
            prevEven = prevEven.next
            if even:
                head = even.next
            else:
                break
        prevOdd.next = evenHead.next
        return oddHead.next
